### ISSUE TITLE
Investigate response times

### ISSUE THREAD

**Issue by aaronschachter** [DoSomething/gambit#704]
Heroku has been sending a number of response time alerts, which seem to have started the [first day we pushed the CampaignBot](https://github.com/DoSomething/gambit/releases/tag/2.5.1) code to production 😿

**Comment by aaronschachter** [comment]
Knocking on wood, but things seem quieter today. As of this morning, we no longer have mData's posting to the legacy `reportback/:campaignName` endpoint (#706) -- and sending members only to the `chatbot` endpoint for Campaigns.

**Comment by deadlybutter** [comment]
Looking over Heroku metrics now,

Quite a few events are closely tied to other actions we've taken,
<img width="372" alt="screen shot 2016-11-02 at 4 39 03 pm" src="https://cloud.githubusercontent.com/assets/897368/19946632/176edd70-a11b-11e6-90e4-c508a9967464.png">

But quite a few aren't (Including the two errors fired this afternoon)
<img width="457" alt="screen shot 2016-11-02 at 4 41 05 pm" src="https://cloud.githubusercontent.com/assets/897368/19946656/2d294952-a11b-11e6-8ca3-ab54a37be96c.png">

It also seems unrelated to throughput or cpu load
<img width="400" alt="screen shot 2016-11-02 at 4 42 50 pm" src="https://cloud.githubusercontent.com/assets/897368/19946722/8361220e-a11b-11e6-9d60-12077560df9f.png">
(notice the throughput on the bottom has bars in places outside the error which are much higher)

Tomorrow I'm going to do some thorough profiling, these are the steps ill take (in my head)

* Create a test dyno formation w/ a custom version of Gambit. In the procfile define the command to start as `node --prof server.js` instead of the current value.
* Generate substantial number of [requests to chatbot](https://github.com/DoSomething/gambit/blob/develop/documentation/endpoints/chatbot.md#chat.
* Pull profile logs from server w/ `heroku run bash --app ${testservername}`
* Process logs into something readable locally `node --prof-process isolate-0xnnnnnnnnnnnn-v8.log > processed.txt`

Hopefully this nets some valuable insight.

More info here from NodeJS docs: 
https://nodejs.org/en/docs/guides/simple-profiling/

**Comment by aaronschachter** [comment]
Thanks @deadlybutter -- looking forward to seeing the results. 

The other requests that will be regular and worth looking into as well is [**POST** `/v1/signups` ](https://github.com/DoSomething/gambit/blob/develop/documentation/endpoints/signups.md)-- Quicksilver posts to this endpoint to trigger sending the CampaignBot `msg_menu_signedup_external` message to a User when they sign up for a CampaignBot-enabled Campaign on the web.

The DonorsChooseBot endpoint `chatbot?bot_type=donorschoose` won't be regularly used -- only when we run Science Sleuth (or any other SMS Game that should post donation as part of the end game).  We are currently posting to this endpoint on prod now though -- not sure when scheduled to close.

**Comment by deadlybutter** [comment]
Good to know!

**Comment by deadlybutter** [comment]
ah crap, let me write this down before i forget. if we still have performance problems we can always try doing a nodejs cluster in the app. it looks like we have the underused resources to make it worth it (ive done this before, not that hard)

**Comment by deadlybutter** [comment]
`loadtest -n 250 -c 25 -k -H "x-gambit-api-key:" -T "application/json" --data '{"phone": "5555555511", "keyword": "slothieboi"}' -m POST http://localhost:5000/v1/chatbot`

https://www.npmjs.com/package/loadtest

**Comment by deadlybutter** [comment]
Only just started but found this, 

```
     41    2.2%    2.4%  RegExp: ^(\\/?|)([\\s\\S]*?)((?:\\.{1\,2}|[^\\/]+?|)(\\.[^.\\/]*|))(?:[\\/]*)$
```

(Every item recorded after thereafter is less than 0.5% usage)

**Comment by deadlybutter** [comment]
So it seems to be referring to this line https://github.com/DoSomething/gambit/blob/32aeaba8f1a306a70a5769ee432c58c11c76db8c/lib/helpers.js#L187

The problem is even when I comment this code out, it consistently remains at 1%

```
     44    1.2%    1.3%  RegExp: ^(\\/?|)([\\s\\S]*?)((?:\\.{1\,2}|[^\\/]+?|)(\\.[^.\\/]*|))(?:[\\/]*)$
```

**Comment by deadlybutter** [comment]
@aaronschachter so this is where we are at a day after the fix we deployed yesterday.
its obviously better but still something seems to be going wrong. Would you prefer I continued auditing or work features?

<img width="1232" alt="screen shot 2016-11-04 at 11 24 36 am" src="https://cloud.githubusercontent.com/assets/897368/20011258/5a133092-a281-11e6-889c-e89942197575.png">

**Comment by aaronschachter** [comment]
I'm curious if the Quicksilver Signup imports triggers these alerts.  We've had stathat on for a few hours now on production for 4 hours, the POST `/v1/signups1` route will query Northstar GET signups then GET users for each signup that Quicksilver sends -- not sure how often it queries for  `/v1/campaigns` endpoints, but just a few times:

<img width="883" alt="screen shot 2016-11-08 at 3 16 21 pm" src="https://cloud.githubusercontent.com/assets/1236811/20121492/a5780baa-a5c6-11e6-858a-ad894df9e2b4.png">

**Comment by deadlybutter** [comment]
Could explain it, my only question would be why it happens some times & not others. (Events don't fire nearly enough for it to be everytime)

**Comment by aaronschachter** [comment]
@deadlybutter Just happened to have logs opened in another email alert, and the niche import was running. The niche import posts Signup IDs to the Gambit `/signups` endpoint -- Gambit queries Northstar for the Signup, then loads the User by the Signup's Northstar UID, in order to send the SMS to the Northstar User's mobile.

You mentioned running tests on the `chatbot` endpoint here https://github.com/DoSomething/gambit/issues/704#issuecomment-257994381 -- was there any insight? It might be worth running same tests on `signups` endpoint for any clues 🔍

**Comment by deadlybutter** [comment]
`niche import` sounds like a bulk operation? In that case we might want to pull that logic you described into a worker process & queue the signups being imported. Would be _fairly_ simple to do if you want me to work on that at some point

**Comment by aaronschachter** [comment]
Yup, it's a bulk operation where Quicksilver is posting Signup events to the Gambit `/signups` endpoint to trigger Signup confirmation SMS messages. Just checked the logs again upon getting an email alert, and I'm beginning to think this could be the culprit.

Have you had any luck running tests on the `chatbot` and `signups` POST requests?

**Comment by aaronschachter** [comment]
Niche import is currently not sending us Users, but we're still seeing alerts today.. so it doesn't seem like it's the bulk import.

**Comment by deadlybutter** [comment]
@aaronschachter Possible there is other things happening, Niche import just one of them?

**Comment by jamjensen** [comment]
@deadlybutter since Niche import does not explain the long response times, there must be other unknown causes. We have no idea if these alerts are a non-issue or are seriously affecting the experience for a lot of users. It'd be super helpful to gain an understanding so that we can confidently close this issue and start on other gambit features. Could you resume auditing and continue to share the results of your tests on this issue? Thank you!  

@sheyd looping you in here. We're constantly receiving Heroku alerts saying response times have exceeded 1250ms. Would greatly appreciate your help if you have any insight into what may be causing longer response times.

**Comment by deadlybutter** [comment]
@jamjensen as I said above, im inclined to believe multiple things could be happening which are causing this. I'll take a deeper look at the logs & try to correlate them with the errors Heroku is reporting. 

In addition the Aaron seeing the link between bulk import & bad response times (which makes sense), I've seen an uptick in failed request emails. Which to me says we have something else happening causing the app to freak out.

**Comment by aaronschachter** [comment]
Digging into some of the transactions, and some of the Northstar Signups query (which proxy to the Phoenix API Signups query) are slow. The middleware waits on the query result to determine which message to send.. which may account for the long middleware response times too.

<img width="741" alt="screen shot 2017-01-25 at 7 08 05 pm" src="https://cloud.githubusercontent.com/assets/1236811/22346619/94bf0918-e3b9-11e6-8149-bcf066650183.png">

We query the Signups index endpoint on every request to always get the current campaign status for the given user. Even if they've signed up already and we have a document, we make the request to get the latest quantity (which we present to the User when we say "we've got you down for [quantity] [nouns] [verbed]". It might be time to reconsider making the query to get latest Signup Status for each Gambit request.

**Comment by aaronschachter** [comment]
My suspicion is that it's the network requests to Northstar and Phoenix that cause our slow response times -- because we don't send a `chatbot` HTTP response back until those requests have finished (sending either the rendered chatbot message, or an error code).

@deadlybutter and I chatted today about always sending a 200 back once we receive the message. Slow Northstar and Phoenix transactions should still show up in New Relic... but this should help avoid sending the timeout error to Mobile Commons, and potentially stop the apdex alerts.

**Comment by aaronschachter** [comment]
Here's an example where it doesn't look like a network request is the issue... which doesn't exactly help troubleshoot.

<img width="1161" alt="screen shot 2017-01-31 at 7 58 17 am" src="https://cloud.githubusercontent.com/assets/1236811/22478599/6d26cdde-e79f-11e6-8541-69963eedb28a.png">


A nice to have here would be adding custom attributes like User ID, Campaign ID, Signup ID, etc into New Relic: https://docs.newrelic.com/docs/agents/manage-apm-agents/agent-data/collect-custom-attributes

**Comment by aaronschachter** [comment]
Phoenix response times are being handled in https://github.com/DoSomething/devops/issues/261. #810 and work done in #849 close this issue on the Gambit side.
